nitro-protocol/contracts/LedgerFinancingApp.sol

```diff
@@ -0,0 +1,151 @@
+        uint128 dpyNum;
+        uint128 dpyDen;
+        // the block number of the latest principal adjustment
+        uint256 blocknumber;
+        // the current principal. Decreases as the serviceProvider earns via the channel.
+        Funds principal;
+        // the total interest collected so far. Strictly increasing.
+        // The value of the intermediary's allocation can never be less than this:
+        // ie, when the intermediary's collectedInterest grows to be equal to their
+        // allocation, the channel is effectively spent and can be concluded.
+        Funds collectedInterest;
+    }
+
+    enum AllocationIndicies {
+        intermediary, // financier: makes initial deposit and earns interest
+        serviceProvider // borrower: recovers service fees from intermediary's deposit
+    }
+
+    function requireStateSupported(
+        FixedPart calldata fixedPart,
+        RecoveredVariablePart[] calldata proof,
+        RecoveredVariablePart calldata candidate
+    ) external view override {
+        if (proof.length == 0) {
+            // unanimous consensus check
+            require(
+                NitroUtils.getClaimedSignersNum(candidate.signedBy) ==
+                    fixedPart.participants.length,
+                '!unanimous; |proof|=0'
+            );
+            return;
+        } else if (proof.length == 1) {
+            // check that proof[0] -> candidate respects the stated interest rate.
+            // Requires:
+            //  - proof state is unanimous
+            //  - candidate state immediately follows proof state (by turnNum)
+            //  - the intermediary has not taken more funds than owed according
+            //    to the interest rate agreement of the channel
+            require(
+                NitroUtils.getClaimedSignersNum(proof[0].signedBy) == 2,
+                '!unanimous proof state'
+            );
+            require(
+                proof[0].variablePart.turnNum + 1 == candidate.variablePart.turnNum,
+                'turn(candidate) != turn(proof)+1'
+            );
+
+            Funds memory outstandingInterest = getOutstandingInterest(
+                abi.decode(proof[0].variablePart.appData, (InterestAppData))
+            );
+            requireOutcomeIsEarned(
+                proof[0].variablePart.outcome,
+                candidate.variablePart.outcome,
+                outstandingInterest
+            );
+        } else {
+            revert('|proof| > 1'); // does it pay to be this terse with revert messages?
+        }
+    }
+
+    // The outstanding interest is calculated based on:
+    //  - the latest consensus principal
+    //  - the channel's interest rate
+    //  - the time elapsed since the last principal adjustment
+    function getOutstandingInterest(InterestAppData memory appData)
+        private
+        view
+        returns (Funds memory)
+    {
+        uint32 numDays = daysSince(appData.blocknumber);
```

Comment:
To check my understanding: if less than 7200 blocks have passed, **no** interest is claimable by Irene?

Reply:
That's right, yes. This *can* be altered by replacing DPY with BPY (block percentage yield). Using BPY was *not* feasible when I had intended to calculate compounding interest (each elapsed block would be another multiplication / division).

Still, moving from DPY from BPY involves a bit some care, since it probably *increases* risk of overflow problems.